arch/common/shared_irq.c

```diff
@@ -17,3 +18,74 @@ void shared_isr(const void *data)
 		}
 	}
 }
+
+#ifdef CONFIG_DYNAMIC_INTERRUPTS
+
+static struct k_spinlock lock;
+
+void z_isr_install(unsigned int irq, void (*routine)(const void *),
+		   const void *param)
+{
+	struct shared_irq_data *irq_data;
+	unsigned int table_idx;
+	int i;
+	k_spinlock_key_t key;
+
+	table_idx = get_sw_isr_table_idx(irq);
+
+	/* check for out of bounds table index */
+	if (table_idx >= CONFIG_NUM_IRQS) {
+		return;
+	}
+
+	irq_data = &_shared_sw_isr_table[table_idx];
+
+	key = k_spin_lock(&lock);
+
+	/* reached the client limit */
+	if (irq_data->client_num == CONFIG_SHARED_IRQ_MAX_NUM_CLIENTS) {
+		goto out_unlock;
```

Comment:
Also, it's a silent failure.  This function doesn't have a way to report error (vs. the static variant, which will straight up fail to build on overflow!).  Might want to __ASSERT() here instead of returning a fake success, so apps have a way to catch the mistake.

Reply:
Fixed. `__ASSERT()` is much better here than the silent approach.